src/display/api.js

```diff
@@ -2650,6 +2686,19 @@ const InternalRenderTask = (function InternalRenderTaskClosure() {
       this._scheduleNextBound = this._scheduleNext.bind(this);
       this._nextBound = this._next.bind(this);
       this._canvas = params.canvasContext.canvas;
+      if (this._canvas) {
+        if (canvasInRendering.has(this._canvas)) {
+          this.capability.reject(
+            new Error(
+              "Cannot use the same canvas during multiple render() operations. " +
+                "Use different canvas or ensure previous operations were " +
+                "cancelled or completed."
+            )
+          );
+        } else {
+          canvasInRendering.add(this._canvas);
+        }
+      }
```

Comment:
Which test-case required moving this code, and is there another way to deal with that?
(Since the commit message says: `- Reject the InternalRenderTask earlier now that InitializeGraphics is waits for another promise.`)

I've not actually tested it, but based on reading the code I'm not entirely convinced that moving this code will actually do the right thing in general.
Currently `initializeGraphics` will throw https://github.com/mozilla/pdf.js/blob/6c39aff3743564f3418ba17344e14819f923a427/src/display/api.js#L2668-L2672 which thus means that we're reaching this catch https://github.com/mozilla/pdf.js/blob/6c39aff3743564f3418ba17344e14819f923a427/src/display/api.js#L1118-L1121 which thus calls https://github.com/mozilla/pdf.js/blob/6c39aff3743564f3418ba17344e14819f923a427/src/display/api.js#L1054-L1081 as expected and all of the required clean-up will thus run.

With your changes here, it doesn't seem like you're actually *guaranteed* that `complete` is ever called correctly!?

Reply:
It was the test that calls render twice. It doesn't seem to be failing anymore. This code is quite the mix of callbacks and old style promises.